sdcm/utils/version_utils.py

```diff
@@ -125,6 +125,142 @@ def parse_scylla_version(version_to_parse: str) -> version.Version:
 
 
+class ComparableScyllaVersion:
+    """Accepts and compares known 'non-semver' and 'semver'-like Scylla versions."""
+
+    def __init__(self, version_string: str):
+        parsed_version = self.parse(version_string)
+        self.v_major = int(parsed_version[0])
+        self.v_minor = int(parsed_version[1])
+        self.v_patch = int(parsed_version[2])
+        self.v_pre_release = parsed_version[3] or ''
+        self.v_build = parsed_version[4] or ''
+
+    @staticmethod
+    def parse(version_string: str):
+        """Parse scylla-binary and scylla-docker-tag versions into a proper semver structure."""
+        # NOTE: remove 'with build-id' part if exists and other possible non-semver parts
+        _scylla_version = (version_string or '').split(" ")[0]
+
+        # NOTE: replace '~' which gets returned by the scylla binary
+        _scylla_version = _scylla_version.replace('~', '-')
+
+        # NOTE: remove docker-specific parts if version is taken from a docker tag
+        _scylla_version = _scylla_version.replace('-aarch64', '')
+        _scylla_version = _scylla_version.replace('-x86_64', '')
+
+        # NOTE: make short scylla version like '5.2' be correct semver string
+        _scylla_version_parts = _scylla_version.split('.')
+        if len(_scylla_version_parts) == 2:
+            _scylla_version = f"{_scylla_version}.0"
+        elif len(_scylla_version_parts) > 2 and re.search(
+                r"\D+", _scylla_version_parts[2].split("-")[0]):
+            _scylla_version = f"{_scylla_version_parts[0]}.{_scylla_version_parts[1]}.0-{_scylla_version_parts[2]}"
+            for part in _scylla_version_parts[3:]:
+                _scylla_version += f".{part}"
+
+        # NOTE: replace '-0' with 'dev-0', '-1' with 'dev-1' and so on
+        #       to match docker and scylla binary version structures correctly.
+        if no_dev_match := re.search(r"(\d+\.\d+\.\d+)(\-\d+)(\.20[0-9]{6}.*)", _scylla_version):
+            _scylla_version = f"{no_dev_match[1]}-dev{no_dev_match[2]}{no_dev_match[3]}"
+
+        # NOTE: replace '.' with '+' symbol between build date and build commit
+        #       to satisfy semver structure
+        if dotted_build_id_match := re.search(r"(.*\.20[0-9]{6})(\.)(\d\w+)", _scylla_version):
+            _scylla_version = f"{dotted_build_id_match[1]}+{dotted_build_id_match[3]}"
+
+        if match := SEMVER_REGEX.match(_scylla_version):
+            return match.groups()
+        raise ValueError(
+            f"Cannot parse provided '{version_string}' scylla_version for the comparison. "
+            f"Transformed scylla_version: {_scylla_version}")
+
+    def __str__(self):
+        result = f"{self.v_major}.{self.v_minor}.{self.v_patch}"
+        if self.v_pre_release:
+            result += f"-{self.v_pre_release}"
+        if self.v_build:
+            result += f"+{self.v_build}"
+        return result
+
+    def _transform_to_comparable(self, other):
```

Comment:
out of curiosity, did you tested this one with `packaging.version.Version`  ?

Reply:
```
>>> from packaging.version import Version
>>> from packaging.version import LegacyVersion
>>> LegacyVersion("5.2.0") > LegacyVersion('5.2.0~dev')
False
>>> Version("5.2.0") > Version('5.2.0~dev')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/kiparis/.local/lib/python3.10/site-packages/packaging/version.py", line 266, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
packaging.version.InvalidVersion: Invalid version: '5.2.0~dev'
```

The introduced here classes return proper `True` result.

Reply:
This works ?


```python
version_utils.ComparableScyllaVersion("5.2.0") < Version("5.3.0)
```

Reply:
> ```python
> ComparableScyllaVersion
> ```

For now it accepts only `str` and the same class type objects.
So, following will work:
```
```